File: 2017-05-18/W3D4/W3D4/ViewController.m
//
//  ViewController.m
//  W3D4
//
//  Created by Jason Liang on 5/18/17.
//  Copyright © 2017 jasonliang. All rights reserved.
//

#import "ViewController.h"
#import "CustomLayout.h"

@interface ViewController () <UICollectionViewDataSource,UICollectionViewDelegate,
    UICollectionViewDelegateFlowLayout>
@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;
@property (nonatomic) UICollectionViewLayout *firstLayout;
@property (nonatomic) CustomLayout *customLayout;
@property (nonatomic) NSInteger itemCount;
@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view, typically from a nib.
  self.collectionView.delegate = self;
  self.collectionView.dataSource = self;
  self.collectionView.allowsMultipleSelection = YES;
  self.firstLayout = self.collectionView.collectionViewLayout;
  self.itemCount = 20;
  self.customLayout = [[CustomLayout alloc] init];
}



- (void)didReceiveMemoryWarning {
  [super didReceiveMemoryWarning];
  // Dispose of any resources that can be recreated.
}

- (IBAction)changeLayout:(id)sender {
  if (self.collectionView.collectionViewLayout == self.firstLayout) {
    [self.collectionView setCollectionViewLayout:self.customLayout animated:YES];
  } else {
    [self.collectionView setCollectionViewLayout:self.firstLayout animated:YES];
  }
}

- (IBAction)addItem:(id)sender {
  
  // version 1
//  [self.collectionView reloadData];
  
  // version 2 - animated
  
  [self.collectionView performBatchUpdates:^{
    self.itemCount -= 1;
    NSIndexPath *indexPath = [NSIndexPath indexPathForItem:self.itemCount-1 inSection:0];
//    [self.collectionView insertItemsAtIndexPaths:@[indexPath]];
    [self.collectionView deleteItemsAtIndexPaths:@[indexPath]];
  } completion:^(BOOL finished) {
    
  }];
  
}

//MARK:

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
  return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
  return self.itemCount;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
  UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"cell" forIndexPath:indexPath];
  cell.backgroundColor = [UIColor greenColor];
  NSArray *selectedIndexPaths = collectionView.indexPathsForSelectedItems;
  if ([selectedIndexPaths containsObject:indexPath]) {
    cell.backgroundColor = [UIColor redColor];
  }
  return cell;
}

//MARK:

- (BOOL)collectionView:(UICollectionView *)collectionView shouldSelectItemAtIndexPath:(NSIndexPath *)indexPath {
  return indexPath.item % 2 == 1;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
  UICollectionViewCell *cell = [collectionView cellForItemAtIndexPath:indexPath];
  cell.backgroundColor = [UIColor redColor];
}

- (void)collectionView:(UICollectionView *)collectionView didDeselectItemAtIndexPath:(NSIndexPath *)indexPath {
  UICollectionViewCell *cell = [collectionView cellForItemAtIndexPath:indexPath];
  cell.backgroundColor = [UIColor purpleColor];
}

//MARK:

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
  return CGSizeMake(50, 50);
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
  return 10;
}

- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
  return 20;
}

@end

File: 2017-11-07/W2D1/W2D1/ViewController.swift
//
//  ViewController.swift
//  W2D1
//
//  Created by Jason Liang on 11/7/17.
//  Copyright © 2017 Jason Liang. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
  @IBOutlet weak var titleLabel: UILabel!
  @IBOutlet weak var imageView: UIImageView!
  
  override func viewDidLoad() {
    super.viewDidLoad()
    titleLabel.text = "Jason"
  }
  
  @IBAction func changeName(_ sender: Any) {
    self.titleLabel.text = "Liang"
  }
  
  @IBAction func importPhoto(_ sender: Any) {
    // init picker
    let picker = UIImagePickerController()
    // customize
    picker.delegate = self
    if TARGET_OS_SIMULATOR == 1 {
      picker.sourceType = .photoLibrary
    } else {
      picker.sourceType = .camera
      picker.cameraDevice = .front
      picker.cameraCaptureMode = .photo
    }
    // present the picker
    self.present(picker, animated: true, completion: nil)
  }
  
  func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
    
    if let image = info[UIImagePickerControllerOriginalImage] as? UIImage {
      self.imageView.image = image
      
      picker.dismiss(animated: true, completion: nil)
    }
    
  }
  
  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
    // Dispose of any resources that can be recreated.
  }


}

